boards: frdm_mcxn947: turn OV7670 into a shield

josuah opened the pull request:
Dependencies:

- https://github.com/zephyrproject-rtos/zephyr/pull/88323

The FRDM-MCXN board is featuring an [Arducam connector](https://docs.arducam.com/DVP-Camera-Module/Arduino-GIGA/Arduino-GIGA/Quick-Start-Guide/) originally designed for the [Arduino GIGA R1](https://docs.arduino.cc/tutorials/giga-r1-wifi/giga-camera/), and spread broadly since then for inexpensive OmniVision and low-power Himax image sensor modules.

![](https://docs.arduino.cc/static/d22f4d902e10938d167ca6b9b7d9206d/a6d36/camera-connector-photo.png)

Other modules than the [Olimex OV7670](https://www.olimex.com/Products/Components/Camera/CAMERA-OV7670/) can be used.

![PXL_20250426_115548902](https://github.com/user-attachments/assets/4c53f591-960d-49e5-b71b-4265cd365467)
![PXL_20250426_115605997](https://github.com/user-attachments/assets/e65ff6d9-a01c-437c-862e-8a9ae0df2acb)
![PXL_20250426_115649291](https://github.com/user-attachments/assets/123a5445-4599-4a93-811b-eed7be3f1b86)


This also proposes an `arducam,dvp-20pin-connector` GPIO nexus:
https://github.com/zephyrproject-rtos/zephyr/pull/89131/files#diff-a39283efe7f5b76e206c8d2b9b65c4b45d130ceb1aa6bb8a0083486793172537

josuah committed:
dts: bindings: gpio: add arducam 20-pin camera connector

Add the 20-pin camera connector used by at least Arducam, Waveshare,
Olimex, Arduino, NXP, ST, Adafruit that connects image sensor module
boards and devkits.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
boards: nxp: frdm_mcxn947: migrate ov7670 to a shield

The OV7670 was defined as part of the FRDM-MCXN947 board attached to the
SmartDMA and I2C peripheral. Migrate the devicetree to a reusable shield
that can be swapped with other image sensors.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Tested with UVC (by cherry-picking commits):

Before:

![mpv-shot0001-before](https://github.com/user-attachments/assets/4740eb21-2c3a-45ce-8c9e-1472d957fe9b)

After:

![mpv-shot0001](https://github.com/user-attachments/assets/4a466221-a04c-483e-87ea-3cfd8ba3fea4)

It does not really look like correct data, but it is showing the same thing before and after, which shows the current PR does not break the sensor operation.

josuah commented on dts/bindings/gpio/arducam,dvp-20pin-connector.yaml:
```diff
@@ -0,0 +1,47 @@
+
+  The following node labels can be used for use in shields:
+
+  - dvp_20pin_interface: the video interface that receives the video
+    feed and use with the zephyr,camera chosen node.
+
+  - dvp_20pin_ep_in: video endpoint of the devkit receiving the video
+    feed (see video-interfaces.yaml).
+
+  - dvp_20pin_ep_out: video endpoint of the camera module to interconnect
+    with this endpoint (see video-interfaces.yaml).
+
+  - dvp_20pin_i2c: the I2C device of the devkit on which the camera
+    module is instantiated.
+
+  - dvp_20pin_gpio: the gpio nexus using the arducam,dvp-20pin-connector
+    that defines the camera 'reset' pin, 'power-down' pin, etc.
```
[Shields](https://docs.zephyrproject.org/latest/hardware/porting/shields.html) does not talk [about a `zephyr_...` prefix](https://github.com/zephyrproject-rtos/zephyr/pull/77854#discussion_r1740830325):

https://github.com/zephyrproject-rtos/zephyr/blob/02f85b253a16f71176d8e4995d801ac464a14633/boards/weact/mini_stm32h743/mini_stm32h743.dts#L243-L253

So I went with `dvp_20pin_<sensorname>` and `dvp_20pin_<feature>`.

Let me know if `zephyr_camera_<...>` is better!

ngphibang commented on boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi:
```diff
@@ -91,40 +101,30 @@ nxp_8080_touch_panel_i2c: &flexcomm2_lpi2c2 {
 
 		port {
-			sdma_ep_in: endpoint {
-				remote-endpoint-label = "ov7670_ep_out";
+			dvp_20pin_ep_in: endpoint {
+				remote-endpoint-label = "dvp_20pin_ep_out";
 			};
 		};
```
This should not be here but in the overlay as it is out of the board. If there are multiple ports / endpoints, you can reserve a port for it as below (but it's not the case here):

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_rt11xx.dtsi#L911-#L934

josuah replied:
I removed the `remote-endpoint-label = "dvp_20pin_ep_out";` and it still works!
Is that what I should remove?

I am not sure how to remove the label itself `dvp_20pin_ep_in:` as the shield cannot know where it should insert itself on every board, the `board.dts` has to provide a label for the shield insertion point?

Thank you for your help as I learn DTS further.

ngphibang reviewed (commented):
Thanks for the added shield.
Did you test the video capture sample ? I think you need to update the `frdm_mcxn947_cpu0.overlay` in the sample as well because the camera is now moved out of the board.

josuah commented:
> Did you test the video capture sample ? I think you need to update the `frdm_mcxn947_cpu0.overlay` in the sample as well

Good point! Now I can remove it as the shield would select the [`zephyr,camera = &dvp_20pin_interface;`](https://github.com/zephyrproject-rtos/zephyr/pull/89131/files#diff-7be4373fb239bd1483933b9916f11421758b2a84f21cbbc932c5328d86757e7aR9).

Trying again after enabling the debug logging:

```
...
[00:00:05.271,000] <dbg> main: main: Got frame 307! size: 19200; timestamp 5271 ms
[00:00:05.286,000] <dbg> main: main: Got frame 308! size: 19200; timestamp 5286 ms
[00:00:05.302,000] <dbg> main: main: Got frame 309! size: 19200; timestamp 5302 ms
[00:00:05.318,000] <dbg> main: main: Got frame 310! size: 19200; timestamp 5318 ms
[00:00:05.334,000] <dbg> main: main: Got frame 311! size: 19200; timestamp 5334 ms
[00:00:05.357,000] <dbg> main: main: Got frame 312! size: 19200; timestamp 5357 ms
[00:00:05.373,000] <dbg> main: main: Got frame 313! size: 19200; timestamp 5373 ms
[00:00:05.388,000] <dbg> main: main: Got frame 314! size: 19200; timestamp 5388 ms
[00:00:05.404,000] <dbg> main: main: Got frame 315! size: 19200; timestamp 5404 ms
[00:00:05.420,000] <dbg> main: main: Got frame 316! size: 19200; timestamp 5420 ms
...
```

josuah commented:
I was hitting this error on CI when `remote-endpoint-label` was removed altogether, so went with a workaround setting it to `"empty"` for when it is not used:
```
devicetree error: 'remote-endpoint-label' is marked as required in 'properties:' in /__w/zephyr/zephyr/dts/bindings/video/nxp,video-smartdma.yaml, but does not appear in <Node /soc/peripheral@50000000/smartdma@33000/video-sdma/port/endpoint in '/__w/zephyr/zephyr/misc/empty_file.c'>
```

ngphibang reviewed (commented):
- I am not sure if this 20-pin shield is generic and not specific only to NXP boards. If it is generic (as currently implemented), it will be similar to the dvp_fpc24_mt9m114 shield used on the mimxrt1064_evk. Could you take a look at
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_rt10xx.dtsi#L885-#L890
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/mimxrt1064_evk/mimxrt1064_evk.dts#L352-#L354
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/shields/dvp_fpc24_mt9m114/dvp_fpc24_mt9m114.overlay

- Could you add an entry for the frdm_mcxn947_cpu0 in the video capture sample README, section "Build and Running" as well ?
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture/README.rst

Thanks

josuah committed:
samples: drivers: video: capture: add FRDM-MCXN947

Add FRDM-MCXN947 to the video capture sample documentation and twister
YAML configuration, using the DVP OV7670 camera module newly introduced.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force-push:

- Added a README entry for the video capture sample, preview soon available [here](https://builds.zephyrproject.io/zephyr/pr/89131/docs/samples/drivers/video/capture/README.html#video-capture)
- Modified the devicetree to use the same declaration style as `mimxrt1064_evk`

> not sure if this 20-pin shield is generic and not specific only to NXP boards

It is the closest thing from a standard for DVP we might have, plenty of [Arducam modules](https://www.uctronics.com/camera-modules/camera-breakout-board.html) available via UCTRONICS, and some 3rd-party too, like the [Olimex one for the OV7670](https://www.olimex.com/Products/Components/Camera/CAMERA-OV7670/).

This will help for testing all these drivers!

ngphibang commented on dts/bindings/gpio/arducam,dvp-20pin-connector.yaml:
```diff
@@ -0,0 +1,47 @@
+  not present, and the pins 17 and 18 are used for the same purpose
+  instead.
+
+  The following node labels can be used for use in shields:
```
`can be used in devicetree:`

ngphibang commented on samples/drivers/video/capture/README.rst:
```diff
@@ -70,6 +73,15 @@ commands:
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/video/capture
+   :board: frdm_mcxn947/mcxn947/cpu0
+   :shield: dvp_20pin_ov7670
```
Could you add the display shield as well so that we have a full camera-display example ?

`:shield: dvp_20pin_ov7670,lcd_par_s035_8080`

ngphibang commented on samples/drivers/video/capture/sample.yaml:
```diff
@@ -10,6 +10,7 @@ tests:
       - platform:mimxrt1064_evk:SHIELD="dvp_fpc24_mt9m114;rk043fn66hs_ctg"
       - platform:mimxrt1170_evk/mimxrt1176/cm7:SHIELD="nxp_btb44_ov5640;rk055hdmipi4ma0"
       - platform:mimxrt1170_evk@B/mimxrt1176/cm7:SHIELD="nxp_btb44_ov5640;rk055hdmipi4ma0"
+      - platform:frdm_mcxn947/mcxn947/cpu0:SHIELD="dvp_20pin_ov7670"
```
`platform:frdm_mcxn947/mcxn947/cpu0:SHIELD="dvp_20pin_ov7670;lcd_par_s035_8080"`

ngphibang commented on dts/bindings/gpio/arducam,dvp-20pin-connector.yaml:
```diff
@@ -0,0 +1,47 @@
+  - dvp_20pin_ep_in: video endpoint of the devkit receiving the video
+    feed (see video-interfaces.yaml).
+
+  - dvp_20pin_ep_out: video endpoint of the camera module to interconnect
```
the output endpoint is not generic as `dvp_20pin_ep_out` but usually a specific camera sensor endpoint such as : `ov7670_ep_out`, so I think we should remove this to avoid confusion ?

josuah replied:
Good point, I forgot to update the `.yaml` file after removing it from the `.dtsi`...

josuah commented on boards/shields/dvp_20pin_ov7670/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_DVP_20PIN_OV7670
+	def_bool $(shields_list_contains,SHIELD_DVP_20PIN_OV7670)
```
I think this is a silent bug as it does not have any effect as not used anywhere, but still a bug!

https://github.com/zephyrproject-rtos/zephyr/blob/8661c3caea02990daeeb06b6e74c7f96c2fc44f3/scripts/kconfig/kconfigfunctions.py#L933-L944

josuah replied:
I also pushed a fix for the `dvp_24fpc_mt9m114` shield, I can move it to a separate PR too if neede.
Thank you for the proofreading!

josuah committed:
boards: shields: dvp_fpc24_mt9m114: search shield in lowercase

The script zephyr/scripts/kconfig/kconfigfunctions.py used by
Kconfig's "shields_list_contains" searches shield names in lower-case,
like they are passed through the SHIELD in CMake, --shield argument
in west. Make the mt9m114 shield upper-case. This has no effect currently
as this Kconfig entry is not used anywhere in Zephyr.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented on include/zephyr/dt-bindings/gpio/dvp-20pin-connector.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef INCLUDE_ZEPHYR_DT_BINDINGS_GPIO_DVP_20PIN_CONNECTOR_H_
+#define INCLUDE_ZEPHYR_DT_BINDINGS_GPIO_DVP_20PIN_CONNECTOR_H_
+
+/** Pin number mask (0..20). */
+#define DVP_20PIN_MASK 0xff
```
I wanted to avoid copy-pasting something without understanding it, but let's use the same thing as everything else: `gpio-map-mask = <0xffffffff 0xffffffc0>;`

josuah commented on dts/bindings/gpio/arducam,dvp-20pin-connector.yaml:
```diff
@@ -0,0 +1,40 @@
+  - dvp_20pin_gpio: the GPIO nexus using the arducam,dvp-20pin-connector
+    that defines the camera pins ('reset' (PEN), 'power-down' (PDN)...)
+
+  See <zephyr/dt-bindings/gpio/dvp-20pin-gpio.h> for pin description.
```
Committed!

josuah commented:
Typo-fixes done and tested again using UVC temporarily cherry-picked on top.

kartben commented on boards/shields/dvp_20pin_ov7670/doc/index.rst:
```diff
@@ -0,0 +1,81 @@
+Requirements
+************
+
+This shield can be used with any board which provides a 18 or 20-pin header spread over two rows
```
```suggestion
This shield can be used with any board that provides an 18 or 20-pin header spread over two rows
```

kartben commented on dts/bindings/gpio/arducam,dvp-20pin-connector.yaml:
```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  GPIO pins exposed on ArduCam 20-pin header camera connector,
+  originally designed to connect Arducam modules to Arduino Giga R1,
```
This makes sure there is a short "title" instead of the otherwise long sentence.
You may also used the new-ish `title` property, actually.

```suggestion
  GPIO pins exposed on ArduCam 20-pin header camera connector.
  Originally designed to connect Arducam modules to Arduino Giga R1,
```

kartben commented on include/zephyr/dt-bindings/gpio/dvp-20pin-connector.h:
```diff
@@ -0,0 +1,32 @@
+ * @name Arducam DVP 20-pin or 18-pin connector pinout
+ * @{
+ */
+#define DVP_20PIN_SCL	3	/**< I2C pin */
+#define DVP_20PIN_SDA	4	/**< I2C pin */
```
nit: explicit better one is clock the other is data?

ngphibang reviewed (approved):
Thanks !
